Golden Images Support For Heterogeneous Clusters (#49)
kubevirt/enhancements: veps/sig-storage/dic-on-heterogeneous-cluster/dic-on-heterogeneous-cluster.md

cluster. Their purpose is to ensure these images automatically available and kept up-to-date. The original design of the
golden images is documented in the [kubevirt/community repository](https://github.com/kubevirt/community/blob/69d061862e0839608932d225a728a7a6e7a89f29/design-proposals/golden-image-delivery-and-update-pipeline.md).

The initial design assumed homogeneous clusters, where all nodes in the cluster share the same architecture. However, as
there is a need to support heterogeneous clusters, where nodes may have different architectures (e.g., `arm64`, `amd64`,
`s390x`), this assumption does not apply anymore, and some changes are required to support this use-case.

This enhancement proposal addresses the challenges of managing golden images in heterogeneous Kubernetes clusters.

## Motivation

The high level flow of the golden images is as follows:

1. The HyperConverged Cluster Operator (HCO) image, includes predefined DataImportCronTemplate files. HCO generates
   a list of `DataImportCronTemplate` objects in the `SSP` CR, based on these files.
2. The SSP operator creates `DataImportCron` CRs from the `DataImportCronTemplate` objects in the `SSP` CR.

> 3. Either the SSP operator or the CDI operator creates a `DataSource` CRs based on the `DataImportCron` CRs.

Reviewer: @fossedihelm Would the templates use a platform specific name for the default datasource in a heterogeneous cluster?

Author: This is part of this suggestion, yes.

Reviewer: Got it, this described in
`For each architecture-specific template, the SSP operator will set the DATA_SOURCE_NAME to the name of the architecture-specific DataSource. The DataSource can be identified using its template.kubevirt.io/architecture and the cdi.kubevirt.io/storage.import.datasource-name labels.`
Is this acceptable for UI?

Author: UI will need to support multiple, arch-specific templates anyway. And this specific change is transparent for the UI - the UI does not modify the `DATA_SOURCE_NAME` parameter.

Author: @metalice - please keep me honest here.

  multiple architecture-specific images). However, when the cdi-importer pulls an image, it selects one suitable for the
  node's architecture, which may not match the architecture of the VM being created. For example, pulling an `arm64`
  image for a VM running on an `amd64` node will cause the VM to fail.

* Users may need to create VMs with specific architectures to run architecture-specific applications. For that, they
  will want to create a custom golden image with a specific architecture, and use it to create VMs with the same
  architecture. This use-case is not supported by the current design.

This enhancement proposal addresses these challenges to support golden images in heterogeneous Kubernetes clusters

## Goals

- Enable users to create virtual machines (VMs) with specific architectures.
- Allow users to define custom golden images with multi-architecture support.
- Ensure architecture-specific `DataSource` CRs are searchable by image architecture.
- Make the golden image resources (e.g. `DataImportCron`, `DataSource`, `DataVolume`, `VolumeSnapshot`, and `PVC`),
  searchable by their architecture.

> - Make the golden image resources, searchable by their data source name (e.g. `centos-stream9`).

Author: I'll drop it.

workload nodes instead.

### Reconciling the `SSP` Custom Resource
When reconciling the `SSP` CR, HCO will:

1. Populate the `spec.cluster.workloadArchitectures` filed in the `SSP` CR with the architectures of the workload nodes
   currently present in the cluster.
2. Add the `ssp.kubevirt.io/dict.architectures` annotation to each **predefined** `DataImportCronTemplate` object in the
 `SSP` CR's `spec.commonTemplates.dataImportCronTemplates` field. This annotation will include a comma-separated list
 of the supported architectures.
   > **Note**: For the **custom** `DataImportCronTemplate` objects, HCO will not add this annotation. Users are
   > responsible for adding the annotation to their custom `DataImportCronTemplate` objects, based on the architectures
   > supported by the specific image.

For example:

```yaml
apiVersion: ssp.kubevirt.io/v1beta3
kind: SSP
spec:
  cluster:
    workloadArchitectures:
      - amd64
      - arm64
      - s390x
  commonTemplates:
    dataImportCronTemplates:
      - metadata:
          name: "centos-stream9-image-cron"
          annotations:

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Reviewer 2: I like this idea, maybe we can get this feature working without involving SSP?

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Reviewer 3: Yes that's why I'm suggesting this, you could even fold the DIC reconciliation into HCO at that point IMHO. It would also allow users to define their own custom DICs without needing to go through HCO and SSP.

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Author: > maybe we can get this feature working without involving SSP?

Almost... we still need to make sure that SSP will deploy arch specific templates, with the right arch apecific DataSource.

> It would also allow users to define their own custom DICs without needing to go through HCO and SSP.

They can already do that, from day one, IIRC.

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Author: > Yes that's why I'm suggesting this, you could even fold the DIC reconciliation into HCO at that point IMHO.

I thought about it, but I'm not sure I agree. In general, HCO purpose is to manage the KubeVirt components, and it was never handle the low level CRs, like VMs, Volumes and so on. I think we should leave it like that, because it a mix of responsibilities. In addition, SSP should still manage the templates to be functional in a multi-arch cluster, and so we still need to inform SSP somehow about it. It doesn't have to be the DICTs as suggested, but it should be done somehow. Bottom line, I'm not totally against it, but I would consider it very carefully.

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Reviewer 4: The DataImportCron API was designed with a single source and a single import DV at a time in mind. Its logic of polling and import was also designed that way. Supporting multi-arch internally by a DataImportCron will require redesigning both API and logic, and keeping backward compatibility will make it quiet clumsy.

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Author: > Supporting multi-arch internally by a DataImportCron will require redesigning both API and logic, and keeping backward compatibility will make it quiet clumsy

This is the situation across all components. But the logic should be done somewhere.

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Reviewer 4: Reading your doc I don't see real API changes (only minor additions) when the multi-arch logic is in HCO+SSP.

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Reviewer 2: > In addition, SSP should still manage the templates to be functional in a multi-arch cluster, and so we still need to inform SSP somehow about it.

Yes, we would need to keep that part (`spec.cluster.workloadArchitectures`) but only for templates. Multi arch images could work without SSP.

>             ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"

Author: @0xFelix - the updated doc contain 2 options about it - one with the new field and one without it. I'll need your input to choose the right one. Please look at the alternatives section

`template.kubevirt.io/architecture` label, the SSP operator will remove it. This behavior addresses two scenarios:

1. Cleaning up resources in the previous format during an upgrade from an older HCO version that does not implement this
   feature.
2. Cleaning up after a user adds the `cdi.kubevirt.io/storage.import.architectures` annotation to their custom
   `DataImportCronTemplate` object.

### Creating and Updating the `DataSource` Custom Resources

> **note**: The `DataSource` CR can be created by either the CDI operator or the SSP operator.

When creating or updating a `DataSource` CR, the `template.kubevirt.io/architecture` and the
`cdi.kubevirt.io/storage.import.datasource-name` labels must be propagated from the corresponding `DataImportCron` CR.

### Loading the Image

When the cdi-importer pulls an image to populate a `VolumeSnapshot` (or the `PVC`), it will use the architecture
specified in the `spec.template.spec.source.registry.platform.architecture` field of the `DataSource` CR, if provided.
The importer will pull the image corresponding to the specified architecture and loads it into the `VolumeSnapshot`.

> > **Note**: the `os` part of the image platform is always `linux`. This refers to the runtime environment of the image,
> > not its content, which may represent different operating systems. On Kubernetes clusters, images always run on the
> > linux OS.

Reviewer 5: > On Kubernetes clusters, images always run on the linux OS.

Kubernetes supports Windows nodes and therefore also Windows containers, with that said, it is true that KubeVirt and its KubeVirt adjacent software only work on Linux nodes. 

So maybe something along the lines of:
```suggestion
> **Note**: The `os` field of the image platform is typically `linux`, reflecting the runtime environment rather than the image contents (i.e., VM images in KubeVirt may represent different operating systems).
```

If the image is not a manifest or does not include the specified architecture, the CDI will report this failure in a
condition on the `DataSource` and `DataImportCron`  CRs.

If the `spec.template.spec.source.registry.platform.architecture` field is not populated, the cdi-importer will pull the 
image without specifying an architecture, maintaining the current behavior.

### Creating a Virtual Machine Templates
The SSP operator will deploy all templates supported in the cluster, corresponding to the architectures listed in the 
`spec.cluster.workloadArchitectures` field of the `SSP` CR. The operator will dynamically add or remove templates if the
architecture list in the SSP CR changes.

For each architecture-specific template, the SSP operator will set the `DATA_SOURCE_NAME` to the name of the
architecture-specific `DataSource`. The `DataSource` can be identified using its `template.kubevirt.io/architecture` and
the `cdi.kubevirt.io/storage.import.datasource-name` labels.

### Creating a Virtual Machine from a Template

> Assuming the user knows the architecture of the VM they want to create, the kuevirt-plugin (UI) will rompt the user to

Reviewer 5: ```suggestion
Assuming the user knows the architecture of the VM they want to create, the kuevirt-plugin (UI) will prompt the user to
```

as the type of the `spec` field is CDI's `DataImportCronSpec`, that one of its nested fields is the
`DataVolumeSourceRegistry` type.

<!--
Tangible API examples used for discussion
-->

## Alternatives

HCO could create a new `DataImportCronTemplate` object for each architecture in the `SSP` CR, and the user will be
responsible for.

pros:

- no changes are required from SSP, except for adopting the new CDI API.

cons:

- the user will be responsible for creating and maintaining the `DataImportCronTemplate` objects, and
  the user will need to create a new `DataImportCronTemplate` object for each architecture.
- the SSP CR will be huge, and will contain a lot of `DataImportCronTemplate` objects. It will be hard to maintain it
  or understand it.
- it looses the meaning of the templating mechanism, by the `DataImportCronTemplate` type.

## Scalability

n/a

## Update/Rollback Compatibility

### Upgrade Compatibility
As outlined above, the SSP operator should maintain the current behavior for `DataImportCronTemplate` objects, that
do not include the `ssp.kubevirt.io/dict.architectures` annotation. Similarly, the cdi-importer preserve its existing

> behavior when the `DataSource`'s `spec.template.spec.source.registry.platform.architecture` is not populate.

Reviewer 5: ```suggestion
behavior when the `DataSource`'s `spec.template.spec.source.registry.platform.architecture` is not populated.
```

> his ensures forward compatibility for all`DataImportCronTemplates` objects. For instance, even after

Reviewer 5: ```suggestion
this ensures forward compatibility for all`DataImportCronTemplates` objects. For instance, even after
```

# VEP #NNNN: Golden Images Support For Heterogeneous Clusters

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [ ] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
- [ ] (R) Target version is explicitly mentioned and approved
- [ ] (R) Graduation criteria filled

## Overview

> Golden images are commonly used disk images that are used to create virtual machines (VMs) in a Kubernetes cluster.

Reviewer 4: OS boot disk images

Their purpose is to ensure these images automatically available and kept up-to-date. The original design of the golden
images is documented in
the [kubevirt/community repository](https://github.com/kubevirt/community/blob/69d061862e0839608932d225a728a7a6e7a89f29/design-proposals/golden-image-delivery-and-update-pipeline.md).

The initial design assumed homogeneous clusters, where all nodes in the cluster share the same architecture. However, as
there is a need to support heterogeneous clusters, where nodes may have different architectures (e.g., `arm64`, `amd64`,
`s390x`), this assumption does not apply anymore, and some changes are required to support this use-case.

This enhancement proposal addresses the challenges of managing golden images in heterogeneous Kubernetes clusters.

## Motivation

The high level flow of the golden images is as follows:

1. The HyperConverged Cluster Operator (HCO) image, includes predefined DataImportCronTemplate files. HCO generates
   a list of `DataImportCronTemplate` objects in the `SSP` CR, based on these files.
2. The SSP operator creates `DataImportCron` CRs from the `DataImportCronTemplate` objects in the `SSP` CR.

> 3. Either the SSP operator or the CDI operator creates a `DataSource` CRs based on the `DataImportCron` CRs.

Reviewer 2: SSP creates DataSources only for its templates IIRC

Author: @0xFelix - could you please check that?

>    The CDI monitors the `DataImportCron` CRs, and ensures the corresponding `DataSource` CRs are updated as needed.

Reviewer 4: The CDI monitors -> CDI controls

Author: I don't this "controls" is better than "monitors" in this context. I mean, CDI watch them and react to changes in them, but perform actions on different object. So I think "monitors" is a better choice here.

Reviewer 4: The CDI DataImportCron controller updates the DataImportCron CR status, so I guess it's more than  monitoring.

> 4. The CDI operator checks the image URL or the ImageStream periodically (according to the cron expression in the

Reviewer 4: The CDI operator checks -> A CDI controller/cronjob polls

>    `DataImportCron` CR), and if the image is updated, it creates a new `VolumeSnapshot` (or a `PVC`), loads the

Reviewer 4: loads -> imports

   latest version of the requested image into this new `VolumeSnapshot`/`PVC`, and modifies the `DataSource` CR to point
   to the latest `VolumeSnapshot`/`PVC`.
5. When creating a VM, users set the VM's `spec.dataVolumeTemplate` field, to point to the desired `DataSource` CR.
   The CDI operator create a new `PVC`, and clones the `VolumeSnapshot`/`PVC` that is referenced by the `DataSource`,
   into this `PVC`.
   > **Note**: Creating a VM is often done using a pre-prepared template, which is created by the SSP operator, and

>    already points to the desired `DataSource` CR.

Reviewer 2: OKD/OCP only though, Templates are not available on Kubernetes

the third aspect: the architectures supported by the image.

### Maintaining a List of Supported Architectures

The HCO will be enhanced to track the architectures of workload nodes in the cluster. It will maintain a distinct list
of these architectures, derived from the `status.nodeInfo.architecture` field of each node.

This list will be dynamically updated as nodes are added or removed from the cluster.

By default, the HCO will identify workload nodes as those labeled with `node-role.kubernetes.io/worker`. If the
`spec.workloads.nodePlacement` field in the `HyperConverged` CR is defined, HCO will use this field to determine the
workload nodes instead.

### Reconciling the `SSP` Custom Resource

When reconciling the `SSP` CR, HCO will:

    1. Populate the `spec.cluster.workloadArchitectures` filed in the `SSP` CR with the supported architectures.
2. Add the `ssp.kubevirt.io/dict.architectures` annotation to each **predefined** `DataImportCronTemplate` object in the
   `SSP` CR's `spec.commonTemplates.dataImportCronTemplates` field. This annotation will include a comma-separated list

>    of the supported architectures.

Reviewer 2: ```suggestion
   of the architectures supported by the registry source.
```

   > supported by the specific image.

For example:

```yaml
apiVersion: ssp.kubevirt.io/v1beta3
kind: SSP
spec:
  cluster:
    workloadArchitectures:
      - amd64
      - arm64
      - s390x
  commonTemplates:
    dataImportCronTemplates:
      - metadata:
          name: "centos-stream9-image-cron"
          annotations:
            ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"
    ...
```

### Reconciling the `DataImportCron` Custom Resources

Upon reconciling the SSP CR, for each `DataImportCronTemplate` object, the SSP operator shall do the following:

1. determine the required architectures. The required architectures are the intersection of
   the values listed in the `spec.cluster.workloadArchitectures` field, and the
   `ssp.kubevirt.io/dict.architectures` annotation. For example, if the
   `ssp.kubevirt.io/dict.architectures` annotation is `"arm64,amd64"` and the
   `spec.cluster.workloadArchitectures` is `["arm64", "s390x"]`, the required architectures are `["arm64"]`.

   > **Note**: for pre-defined `DataImportCronTemplate` objects, the `ssp.kubevirt.io/dict.architectures`

>    annotation is expected to be the same as the architectures listed in the `spec.cluster.workloadArchitectures` field.

Reviewer 2: IMO a DICT should have all the arches it supports in it's annotation, regardless of the architectures available in the cluster right now.

Reviewer 2: HCO does not need to know it, but someone supplies this information to HCO (the DICTs) and they should know.

Author: The DICT are based on hard coded files in the HCO image. DICT objects in these files are referring to images in a registry (or in an imageStream). These images are frequently changing - this is why we have the DIC mechanism. 

So I can assume that all the images are actually manifests, that points to a fixed list of arch specific images, and only the specific images are changed. But who can promise that this assumption is valid? The DICT file created at image build time. The image manifest is changing all the time. In theory, it can become a regular image, or maybe the arch list in the manifest will be changed. This is not something HCO can monitor.

   The reason for this behavior is for custom `DataImportCronTemplate`, where the user may want to set the architectures
   to match the image manifest on one hand, and to avoid creating `DataImportCron` CRs for architectures that are not
   existing in the cluster.
2. For each required architecture, the SSP operator shall create or update a `DataImportCron` CR. For each one of these

>    CRs, the SSP operator shall:

Reviewer 2: How should it name the DICs? Append suffix to name of the DICT?

Author: I don't think we need to mandate it. Personally I don't like to use name of things for the software logic. That why I added the labels for.

I wrote it in line 185:
> **Note**: The name of the `DataImportCron` CR is out of scope of this document, but it must be unique, and to reflect both the `DataImportCronTemplate` name, and the specific architecture; e.g. `centos-stream9-image-cron-amd64` or `fedora-image-cron-arm64`

### Creating a Virtual Machine from a Template
Assuming the user knows the architecture of the VM they want to create, the kuevirt-plugin (UI) will rompt the user to
select the desired architecture. Once selected, the UI will display templates corresponding to the chosen architecture.

While the above assumption is valid, this approach contrasts with practices observed in other virtualization solutions.
For instance, AWS integrates architecture selection directly into the template selection UI:

![aws-select-arch-ui.png](aws-select-arch-ui.png)

## API Changes

### CDI API Changes
The following API changes are required:
The `DataVolumeSourceRegistry` type in the CDI API will be extended with the new `platform` field, with the new type of
`PlatformOptions`; e.g.

```go
type DataVolumeSourceRegistry struct {
//Platform describes the minimum runtime requirements of the image
// +optional
Platform *PlatformOptions `json:"platform,omitempty"`
}
```

```go
type PlatformOptions struct {
//Architecture specifies the image target CPU architecture
// +optional
Architecture string `json:"architecture,omitempty"`
// +optional
//Variant specifies the image target CPU architecture variant

> Variant string `json:"variant,omitempty"`

Reviewer 2: i.e. something like armv7 or armv8?

Reviewer 5: More so armv7-a / armv7-r to avoid potential binary compatibility issues.

Author: This is part of CDI POC, so I left it there. But at this phase, I don't think we need it in the e2e feature.

Reviewer 4: @nunnatsa not sure we need this api change if we keep the import-by-digest concept, as each arch has its own digest in the manifest.

Reviewer 5: On second thought since the doc states that variant is out-of-scope for this proposal, perhaps it'd be a good idea to omit this API field after all.

Reviewer 2: If it is out of scope let's drop this

do not include the `ssp.kubevirt.io/dict.architectures` annotation. Similarly, the cdi-importer preserve its existing
behavior when the `DataSource`'s `spec.template.spec.source.registry.platform.architecture` is not populate.

his ensures forward compatibility for all`DataImportCronTemplates` objects. For instance, even after
upgrading the HCO to a version that implements this feature, the custom `DataImportCronTemplate` objects will remain
unaltered and continue to function as expected, allowing users to update them at their own pace.

### Rollback Compatibility
For the same reason, an older version of the HCO will reconcile the `DataImportCronTemplate` objects in the `SSP` CR
without including the `spec.workloads.nodePlacement` field and the `ssp.kubevirt.io/dict.architectures` annotation.
The SSP operator will treat these as legacy `DataImportCronTemplate` objects, as described above.

The SSP operator is responsible for removing any `DataImportCron` CRs created by the SSP that are not defined by any
current `DataImportCronTemplate` object in the SSP CR.

## Functional Testing Approach

> Effective functional tests will require to build a test environment with heterogeneous cluster, which presents a

Reviewer 2: Depending on how CDI implements the import feature we could import multiple arches on a single arch cluster

Author: That will work for ssp and cdi test, but not for e2e with hco.

## Motivation

The high level flow of the golden images is as follows:

1. The HyperConverged Cluster Operator (HCO) image, includes predefined DataImportCronTemplate files. HCO generates
   a list of `DataImportCronTemplate` objects in the `SSP` CR, based on these files.
2. The SSP operator creates `DataImportCron` CRs from the `DataImportCronTemplate` objects in the `SSP` CR.
3. Either the SSP operator or the CDI operator creates a `DataSource` CRs based on the `DataImportCron` CRs.
   The CDI monitors the `DataImportCron` CRs, and ensures the corresponding `DataSource` CRs are updated as needed.
4. The CDI operator checks the image URL or the ImageStream periodically (according to the cron expression in the
   `DataImportCron` CR), and if the image is updated, it creates a new `VolumeSnapshot` (or a `PVC`), imports the
   latest version of the requested image into this new `VolumeSnapshot`/`PVC`, and modifies the `DataSource` CR to point
   to the latest `VolumeSnapshot`/`PVC`.
5. When creating a VM, users set the VM's `spec.dataVolumeTemplate` field, to point to the desired `DataSource` CR.

>    The CDI operator create a new `PVC`, and clones the `VolumeSnapshot`/`PVC` that is referenced by the `DataSource`,

Reviewer 4: "The CDI Operator creates" -> "CDI creates"

For example:

```yaml
apiVersion: ssp.kubevirt.io/v1beta3
kind: SSP
spec:
  cluster:
    workloadArchitectures:
      - amd64
      - arm64
      - s390x
  commonTemplates:
    dataImportCronTemplates:
      - metadata:
          name: "centos-stream9-image-cron"
          annotations:
            ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"
    ...
```

### Reconciling the `DataImportCron` Custom Resources

Upon reconciling the SSP CR, for each `DataImportCronTemplate` object, the SSP operator shall do the following:

1. determine the required architectures. The required architectures are the intersection of
   the values listed in the `spec.cluster.workloadArchitectures` field, and the
   `ssp.kubevirt.io/dict.architectures` annotation. For example, if the
   `ssp.kubevirt.io/dict.architectures` annotation is `"arm64,amd64"` and the
   `spec.cluster.workloadArchitectures` is `["arm64", "s390x"]`, the required architectures are `["arm64"]`.

   > **Note**: for predefined `DataImportCronTemplate` objects, the `ssp.kubevirt.io/dict.architectures`
   annotation is expected to be the same as the architectures listed in the `spec.cluster.workloadArchitectures` field.

>    The reason for this behavior is for custom `DataImportCronTemplate`, where the user may want to set the architectures

Reviewer 6: Wouldn't be better if HCO rejects custom DIC template, when the user specifies wrong arch, which is not available at the moment?

Reviewer 5: What's the incentive of rejecting custom DIC templates if the user specifies a wrong arch? What if as part of their cluster deployment flow they prefer to setup virt-related operators and resources followed by node expansion which have the relevant hardware?

IMHO at the very least, the import process should be hardware agnostic, that is, it should be possible to create an arm64 image in an homogeneous x86_64 cluster.

Author: > IMHO at the very least, the import process should be hardware agnostic, that is, it should be possible to create an arm64 image in an homogeneous x86_64 cluster.

These images are not small, and we usually even have several versions of the same image before they get garbage collected. In addition, unlike an image registry, we just load these images into a VolumeSnapshot (or a PVC), and we can't reuse image layers. So the storage impact may be meaningful.

I would prefer to not import images if they are not needed.

Reviewer 2: > IMHO at the very least, the import process should be hardware agnostic, that is, it should be possible to create an arm64 image in an homogeneous x86_64 cluster.

That's going to be an issue when using the `pullMethod: node` to make use of a node's registry credentials. For `pullMethod: pod` this should be fairly easy to accomplish.

Reviewer 2: > IMHO at the very least, the import process should be hardware agnostic, that is, it should be possible to create an arm64 image in an homogeneous x86_64 cluster.

For `pullMethod: pod` this should be easy to accomplish but for `pullMethod: node` this is going to be an issue

Reviewer 5: > > IMHO at the very least, the import process should be hardware agnostic, that is, it should be possible to create an arm64 image in an homogeneous x86_64 cluster.
> 
> For `pullMethod: pod` this should be easy to accomplish but for `pullMethod: node` this is going to be an issue

This is a good point that warrants further discussion.
Since `pullMethod: node` is an exception rather than standard behavior, adding a node-selector to the importer spec would be a reasonable solution for this scenario (albeit not a hardware agnostic one). As an alternative, we could inspect the manifest to determine the exact architecture-specific digest (a bit sneaky), though I have concerns about introducing this sort of additional logic to the controller.

Reviewer 7: Actually, when it comes to openshift, those are almost the rule rather than the exception.
But yeah unfortunately I don't see any other way other than making a call against the manifest endpoint and traversing through for the right variant (how bad is this really?)

I really wish the ImageVolumeSource exposed platform.
/cc @Barakmor1
won't we have a similar issue with containerdisk VMs?

Author: @akalenyu, there is a [POC](https://github.com/kubevirt/containerized-data-importer/commit/dbed037c8e337f3ea7a6001cdbec2b9f8106f9ea#diff-89de25cd054c9bd6c5b173614719a25a8b199c1c0aaf845f6c24582062c62a01) by @Acedus, that adds the platform field to the DV API. 

The VEP mentions this new platform field. The VEP assumes it's there.

Reviewer 7: I know, this conversation is about making that API work `pullMethod: node` DVs

Author: So at least podman pulls a new image if the architecture is different.

For example, linux/amd64 quay.io/kubevirt/hyperconverged-cluster-operator@sha256:588796159879eb030bf4e79c56e0c90a6eae4742170ca467bd83d7b37c79a8f9 is a multi-arch manifest image:
```
$ podman pull --platform=linux/amd64 quay.io/kubevirt/hyperconverged-cluster-operator@sha256:588796159879eb030bf4e79c56e0c90a6eae4742170ca467bd83d7b37c79a8f9 
Trying to pull quay.io/kubevirt/hyperconverged-cluster-operator@sha256:588796159879eb030bf4e79c56e0c90a6eae4742170ca467bd83d7b37c79a8f9...
Getting image source signatures
Copying blob sha256:795ac0bcfa0636ba84d382284ccb7cfdb1a6c4b564b95e854b032a5ef586b9f1
Copying blob sha256:063a16c022ba3f913fbd2475bdce3195b26e82805b1357c7f33890a5bab0fdcc
Copying blob sha256:5ac13be8a1b22801d1e75c5307dc934f809801b5da6af520362f50226fb96060
Copying blob sha256:b9435916f7e7425d37641a4adf3d73cba4aedd38d43512500f1ce8e42722dfff
Copying blob sha256:ce9573ae331f91f7e0010e52d0b5dbeb21bdcb5180a841219968a9cf2bf2b674
Copying config sha256:e8fed7a469ee02a761bf53948699aef492e9416177d6dd68830fb5fc1ef0797f
Writing manifest to image destination
e8fed7a469ee02a761bf53948699aef492e9416177d6dd68830fb5fc1ef0797f

$ podman images -f label=app=hyperconverged-cluster-operator
REPOSITORY                                        TAG         IMAGE ID      CREATED     SIZE
quay.io/kubevirt/hyperconverged-cluster-operator  <none>      e8fed7a469ee  6 days ago  192 MB

$ podman pull --platform=linux/arm64 quay.io/kubevirt/hyperconverged-cluster-operator@sha256:588796159879eb030bf4e79c56e0c90a6eae4742170ca467bd83d7b37c79a8f9 
Trying to pull quay.io/kubevirt/hyperconverged-cluster-operator@sha256:588796159879eb030bf4e79c56e0c90a6eae4742170ca467bd83d7b37c79a8f9...
Getting image source signatures
Copying blob sha256:795ac0bcfa0636ba84d382284ccb7cfdb1a6c4b564b95e854b032a5ef586b9f1
Copying blob sha256:8902a2f215fcdfa5f84666a47ca55d1284e67385acffe88d8ea76fbdd554fcc5
Copying blob sha256:0790c5901803c98d83971ddfb00ff77badd3bd7fc73380e1c05cfa2213bb85d9
Copying blob sha256:5b8d048955d9f3216bf3c8b7bb9986d86e43a421439393be1791218dd5a8f383
Copying blob sha256:d653033ce8bf637607650c05db5d695ee44c416796f4c0e0ccda4e2313d15a54
Copying config sha256:df3e2d7e27fe60eec4e1a5c7eec5e5339ef628937d98eec3ca2133df35104063
Writing manifest to image destination
df3e2d7e27fe60eec4e1a5c7eec5e5339ef628937d98eec3ca2133df35104063

$ podman images -f label=app=hyperconverged-cluster-operator
REPOSITORY                                        TAG         IMAGE ID      CREATED     SIZE
quay.io/kubevirt/hyperconverged-cluster-operator  <none>      df3e2d7e27fe  6 days ago  193 MB  <<== the new arm64 image
<none>                                            <none>      e8fed7a469ee  6 days ago  192 MB  <<== amd64 image became dangling
```

Author: I think I  finally got you all...

So using arch specific digest won't work either. Even using cp to copy files from the init container will probably fail, as cp is also an arch specific binary.

It seemd to me that the best option is to add a node selector with the required arch. 

As for rejecting such dics from the webhook - that will enforce cdi to read the node list, and we want to avoid this, and it also breaks eventually consistent.

We can maybe add a condition to the dic saying it can't import images. But again, this will force cdi to know the cluster internals.

HCO can prevent the creation of such dics, but it won't help for independent dics created directly by the user.

Reviewer 5: > making a call against the manifest endpoint and traversing through for the right variant (how bad is this really?)

One problem I can think of is the matter of authorization. A user can specify `pullMethod: node` and bind their own custom pull-secret to the default SA of their namespace which in result would allow the image to be pulled successfully but the CDI controller wouldn't be able to probe it in advance.

There can be more unorthodox alternatives such as forcing an image refresh by using an init-container that pulls the image in advance, but those come at a price.

Reviewer 8: > won't we have a similar issue with containerdisk VMs?

For container disks, we use a label selector for the architecture, which is simple because it runs inside the virt-launcher pod. For PVCs, I guess it’s a bit more complex.

   to match the image manifest on one hand, and to avoid creating `DataImportCron` CRs for architectures that are not
   existing in the cluster.
2. For each required architecture, the SSP operator shall create or update a `DataImportCron` CR. For each one of these
   CRs, the SSP operator shall:
    1. Set the `spec.template.spec.source.registry.platform.architecture` field in the `DataImportCron` CR, to the
       architecture of the image. This field is used by the cdi-importer to pull the desired image from the registry.

>     2. Add the `template.kubevirt.io/architecture` label with the required architecture.

Reviewer 4: why both are needed and not only in the DV template? we should consider whether annotation/field should be added there.

Author: Because it makes all the underline CRs, mostly the DataSource, to be searchable by arch and image name (w/o the arch suffix). It is useful for SSP to set the DataSource name in the template, and for the UI to build a proper display of templates.

Author: Not sure what duplication you refer to.

Anyway, labels are the straight forward way to filter resources in k8s.

Reviewer 4: Specifying the same architecture & datasource values both in the DataImportCron spec and in its labels is duplication, isn't it?

Author: I don't think so. labels are labels and are used for filtering. Spec fields are for the logic. And there are many examples in K8s for similar duplication. 

What is the concern?

Reviewer 4: Ok, got your point.

>     3. Add the `cdi.kubevirt.io/storage.import.datasource-name` label with the value of `DataImportCronTemplate`'s

Reviewer 4: why needed?

Author: Same

            ssp.kubevirt.io/dict.architectures: "arm64,amd64,s390x"
          name: centos-stream9-image-cron
        spec:
          garbageCollect: Outdated
          managedDataSource: centos-stream9
          schedule: "0 */12 * * *"
          template:
            spec:
              source:
                registry:
                  url: docker://quay.io/containerdisks/centos-stream:9
              storage:
                resources:
                  requests:
                    storage: 10Gi
```

For the above `DataImportCronTemplate` object, SSP will create three `DataImportCron` CRs, one for `amd64`,
one for `arm64`, and one for `s390x`. The `DataImportCron` CR for the `amd64` architecture, will look like this:

```yaml
apiVersion: cdi.kubevirt.io/v1beta1
kind: DataImportCron
metadata:
  annotations:
    cdi.kubevirt.io/storage.bind.immediate.requested: "true"
  labels:
    template.kubevirt.io/architecture: "amd64"
    cdi.kubevirt.io/storage.import.datasource-name: "centos-stream9"
  name: centos-stream9-image-cron-amd64
spec:
  garbageCollect: Outdated
  managedDataSource: centos-stream9
  schedule: 0 */12 * * *
  template:
    spec:
      source:
        registry:

>           url: docker://quay.io/containerdisks/centos-stream:9

Reviewer 9: @nunnatsa Why do you want to set a specific architecture in each object, if CDI's importer pod can just check the supported architectures by using this URL? you can get a list of the supported architectures from the image index:
```
╰─$ podman manifest inspect quay.io/containerdisks/fedora:41
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
    "manifests": [
        {
            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "size": 429,
            "digest": "sha256:90de00c08bc09971e00ca5801161943f29e07be8f5e9cd8cc4ce4de9a166db82",
            "platform": {
                "architecture": "arm64",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "size": 429,
            "digest": "sha256:d6f7967bcc9b74112299a6c89e8c76430824ccab1cb9cd486b9b83472aba1351",
            "platform": {
                "architecture": "s390x",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "size": 429,
            "digest": "sha256:ae48475e213bd3d34415e8f123f5a79166b1eb142fdf851654e3adb9d02fb73f",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        }
    ]
}
```

Reviewer 5: Hi @codingben, could you clarify your question? I'm not sure I follow.

Author: @codingben - CDI know nothing about the cluster architectures, nodes and so on, and it won't know in the future. That why this information should be passed somehow, from HCO.

Reviewer 9: @nunnatsa Ack.

          platform:

>             architecture: "amd64"

Reviewer 4: Since the `DataImportCron` controller creates the `DataVolume` based on the template, can't it simply use the `template.kubevirt.io/architecture` instead of changing the `DataVolumeSourceRegistry` API? afaiu DVs can still use the url@digest with multi-arch (each arch has its own digest in the manifest) so no change will need to be done in the importer itself.

Reviewer 4: `DataImportCron` import DVs are using the source digest which is either polled from the `containerImage` manifest by the `cdi-source-update-poller` cronJob, or by the controller from the `imageStream.Status.Tags` in case of `imageStream`. It would be nice to keep the import flow as-is, and only adapt the polling to be arch-aware (if needed and if possible).

Reviewer 5: What about users that use `DataImportCrons` and would like to create standalone DVs but still require a specific architecture?

Author: I can't see how it will work. First, you probably can't do arch-specific polling. Assuming the image listed in the url or in the ImageStream is multi-arch - or actually a manifest, then this is the information CDI got, and this image is what CDI can poll. Also, my assumption is that upon releasing of new version of the image, it built for each arch, and then the manifest is built and pushed to the registry. In another words, we can assume that if there is a new version of the image, this is the case for all the architectures.

Reviewer 4: They can either use the specific arch digest or a specific tag.

Author: > so no change will need to be done in the importer itself.

I can't see how. This is the heart of the change we need to do. The importer **must** be architecture aware - we can't escape from it, no matter what implementation we'll choose.

Reviewer 4: > I can't see how. This is the heart of the change we need to do. The importer must be architecture aware - we can't escape from it, no matter what implementation we'll choose.

why? In a manifest, in addition to the common digest, each arch has its own digest.

Author: @arnongilboa - There is a POC by @Acedus to import arch-specific image, [here](https://github.com/kubevirt/containerized-data-importer/commit/dbed037c8e337f3ea7a6001cdbec2b9f8106f9ea#diff-89de25cd054c9bd6c5b173614719a25a8b199c1c0aaf845f6c24582062c62a01).

This enhancement is actually done on top of this POC. Without this new feature, the implementation will be much complex in CDI, requiring the poller to read the manifest and create the DV with the arch specific image digest. 

But apparently, there is an option in the golang package used to import the image, to specify the architecture - that makes everything much simpler. I'll add this a a bit more background to the VEP.

Reviewer 5: > They can either use the specific arch digest or a specific tag.

Doesn't it kind of defeat the purpose of multi-arch images, though? The user experience would feel a bit awkward in that scenario IMO (that is to say, a user can do that right now). 

I'm not entirely against the idea though, it'd a solid alternative, just need to see if it'd be practical in execution.

`DataImportCronTemplate` objects to the new design at their own pace.
Additionally, it allows for independent implementation of components. For instance, if SSP implements this
design before the HCO, it will not introduce breaking changes.

If a `DataImportCronTemplate` object contains the `ssp.kubevirt.io/dict.architectures` annotation, but
a `DataImportCron` exists with the same name as the `DataImportCronTemplate`, and lacks the
`template.kubevirt.io/architecture` label, SSP will remove it. This behavior addresses two scenarios:

1. Cleaning up resources in the previous format during an upgrade from an older HCO version that does not implement this
   feature.
2. Cleaning up after a user adds the `cdi.kubevirt.io/storage.import.architectures` annotation to their custom
   `DataImportCronTemplate` object.

### Creating and Updating the `DataSource` Custom Resources

> **note**: The `DataSource` CR can be created by either CDI or SSP.

When creating or updating a `DataSource` CR, the `template.kubevirt.io/architecture` and the
`cdi.kubevirt.io/storage.import.datasource-name` labels must be propagated from the corresponding `DataImportCron` CR.

### Loading the Image

> When the cdi-importer pulls an image to populate a `VolumeSnapshot` (or the `PVC`), it will use the architecture

Reviewer 4: pulls -> creates a DataVolume importing

> specified in the `spec.template.spec.source.registry.platform.architecture` field of the `DataSource` CR, if provided.

Reviewer 4: `DataSource` CR -> `DataImportCron` CR

name, e.g. `cdi.kubevirt.io/dict.architectures`) annotation will be copied to the `DataImportCron` CR, as any other
annotation.

CDI will create multiple `DataSource` CRs, one for each architecture, according to the
`cdi.kubevirt.io/dict.architectures` annotation, and will import the images into architecture specific `VolumeSnapshot`
resources. 

When performing the actual image import, if the `cdi.kubevirt.io/dict.architectures` annotation is set in
the `DataImportCron` CR, the CDI will create architecture specific `DataVolume` CR, for each architecture
listed in the annotation. CDI will set the `DataVolome` CR's `spec.source.registry.platform.architecture` field, to the
required architecture.

CDI will add the `template.kubevirt.io/architecture` label to the `DataSource` CR, with the architecture of the image,
and will set the `cdi.kubevirt.io/storage.import.datasource-name` label to the value of the `spec.managedDataSource` 
field in the `DataImportCron` CR.

CDI will need to remove any orphan `DataSource` or `VolumeSnapshot` CRs, that are not referenced by any `DataImportCron`

> CR.

Author: Thank for the comment @aglitke.

First, CDI does not monitor the cluster for node changes, but HCO does it, and pass the information to CDI.

> Asking it to manage a dynamically changing set of images based on nodes entering and leaving the cluster is outside of its scope

I think each component that creates objects in the system, is responsible to the cleanup. So if CDI is creating the DataSource and the VolumeSnapshot resources, it makes sense to me that it should also delete them if needed.

SSP will still need to create architecture specific templates, and will still need to set the `DataSource` name 
in each template, to the architecture specific `DataSource` CR name.

Open questions:
1. SSP currently creates the `DataSource` CRs in some cases. We have two options here:
   1. SSP will have to have the same logic as in CDI, when creating the architecture specific `DataSource` CRs.

>    2. SSP will no longer create the `DataSource` CRs, and will let CDI to do that.

Reviewer 2: SSP creates DataSources for all templates it deploys. If we set the default name of the DataSources in the templates accordingly to the logic that CDI will be using for associating a DataImportCron with a DataSource we should be fine.

Author: My concern is that eventually either SSP or CDI will change the implementation a bit, and they won't match anymore.

2. What is the best way for SSP to find out which architecture specific templates to create? 

>    1. SSP can use the new `spec.cluster.workloadArchitectures` field in the `SSP` CR, and create the templates

Reviewer 2: We have other Templates without a default DataSource (Windows), to me that's acceptable. For most templates we provide we should have a fitting disk image anyway.

Reviewer 2: That leaves it open to a user to provide their own image. Keep in mind that DataSources in Templates are not hardcoded, they can be provided as a parameter.

Reviewer 6: Btw the default DataSurces for windows exists (they are created by SSP), we just don't provide images for them (it is user responsibility).
The default `DATA_SOURCE_NAME` is hardcoded in templates (e.g. https://github.com/kubevirt/common-templates/blob/master/templates/windows11.tpl.yaml#L202), it is not set by ssp operator.

Author: > The default DATA_SOURCE_NAME parameter is hardcoded in templates (e.g. https://github.com/kubevirt/common-templates/blob/master/templates/windows11.tpl.yaml#L202), it is not set by ssp operator.

This is one of the thing we need to change as part of the implementation of this VET, as you mentioned at https://github.com/kubevirt/common-templates/issues/680. And it's also detailed in the VET, in the "Creating a Virtual Machine Templates" section.

   
   2. SSP can use the `cdi.kubevirt.io/dict.architectures` annotation in each `DataImportCronTemplate`, and create the
      templates accordingly.
   
      The problem with this approach is that SSP will need to implement the logic to find out
      which `DataImportCronTemplate` objects is related to each template. SSP can use the `spec.managedDataSource` field
      in the `DataImportCronTemplate` object, to find the templates using the same `DataSource` name. It may be too 
      complex to implement this solution.
      > **Important**: this alternative will require no API changes in SSP, as the `spec.cluster.workloadArchitectures`
        field is not needed.

pros:

- no changes are required from SSP, except for adopting the new CDI API, and changes regarding the VM templates.
- The development of the changes in SSP and in HCO is not dependent on CDI.
- If the 2nd option in the 2nd open question is chosen, the development of the changes in HCO is not dependent on SSP,
  and each component can be developed and released independently.
- In deployments without SSP, this solution will work as well, if the user creates the `DataImportCron` CRs manually.

cons:

> - template handling in SSP seems to be more complex.

Reviewer 2: I don't think so, we can create templates without boot sources, that's fine.

I prefer this alternative, it requires less changes in SSP and as you pointed out it should work well in deployments without SSP too.

Author: @akalenyu - could you please comment on this?

Reviewer 7: > AFAIK, there are CDI users, where it deployed w/o SSP/HCO, that do use DataImportCron. Shouldn't we support heterogeneous clusters for these users as well?

once registry DVs support specifying OCI platform fields
https://github.com/opencontainers/image-spec/blob/c05acf7eb327dae4704a4efe01253a0e60af6b34/specs-go/v1/descriptor.go#L53
the building blocks for heterogeneous clusters do exist, so why would it not be supported? 

I agree with Adam that this overlaying logic would not belong in sort of a "core" project

Reviewer 2: I lost track a bit, so we're back to creating multiple DSes for each architecture and SSP will fill in the DS names into the Templates it creates? That would make sense to me as well.

Author: Creating one DICT by HCO ==> multiple DICs by SSP

`ssp.kubevirt.io/dict.architectures` annotation to the `DataImportCronTemplate` object in the `SSP` CR, only with the
architectures that are supported by the image, and by the cluster.

However, these images are frequently updated, and we can never know if the architecture list that was set at HCO image
build time, is still valid or not, at any point in the future.

It seems like a low risk, but it will require a new release of HCO to update the architecture list.

## Scalability

n/a

## Update/Rollback Compatibility

### Upgrade Compatibility
As outlined above, SSP should maintain the current behavior for `DataImportCronTemplate` objects, that
do not include the `ssp.kubevirt.io/dict.architectures` annotation. Similarly, the cdi-importer preserve its existing
behavior when the `DataSource`'s `spec.template.spec.source.registry.platform.architecture` is not populated.

This ensures forward compatibility for all`DataImportCronTemplates` objects. For instance, even after
upgrading the HCO to a version that implements this feature, the custom `DataImportCronTemplate` objects will remain

> unaltered and continue to function as expected, allowing users to update them at their own pace.

Reviewer 10: From what I gather, the DataSource names for our default golden images change after this multi arch image support feature is turned on. That impacts user's existing investment in the default golden images.

Is there a path we can take that allows people to enable this feature while not making them do any "work" to adapt their current gitops flows simply to make their current yamls continue to work?

Author: Thanks for the important input, @davidvossel.

Could you please add more details? For the predefined (common) DICTs, and their corresponding templates, what manipulations did you see? I mean, do they change the DataSource name (or the `DATA_SOURCE_NAME` parameter)? how? can you please share an example?

Reviewer 6: Not directly connected to David's comment, but we will have to change source DS names in templates too https://github.com/kubevirt/common-templates/issues/680

Reviewer 6: Yes, but we still have to solve it for common templates itself. And wouldn't be better if the arch is already part of templates and then SSP wouldn't have to do anything with the `DATA_SOURCE_NAME`? I don't expect the DATA_SOURCE_NAME would change during the time.

Author: OK. I don't like using names for logic. I really prefer to use solid data like labels in this case. But I can see how setting naming standard can ease things. I can live with it...

Reviewer 2: > Is there a path we can take that allows people to enable this feature while not making them do any "work" to adapt their current gitops flows simply to make their current yamls continue to work?

On a homogeneous cluster we could keep the DataSource names as is. On heterogeneous clusters though the name needs to be changed to avoid ambiguity.

Author: OK. So I  have a cluster and I decided to extand it. I've got atractive offer to purchase new machines, but with different architecture. So now my homogeneous cluster became heterogeneous. Should we keep the existing dataSources/DataImportCron/templates, or replace them with new arch specific ones?

Reviewer 2: Can we make a DataSource arch-aware so it can point at multiple PVCs for different arches?

Author: I don't think so. I think @arnongilboa can explain better than I, but in general, IIUC, it requires meaningful not backward compatible apu changes.

Reviewer 2: > So, with regards to seamless updates, what can be done to reduce disruption for people who are already using the existing DataSources by name?

So this is why I was hinting at making DataSources arch-aware, so the current names can be kept. But @aglitke is right that this is a mix of concerns.

A better alternative might be to keep a DataSource with the old name that points to an image with the same arch as the control plane architecture? This should allow for somewhat backward-compatible behavior?

```yaml
apiVersion: cdi.kubevirt.io/v1beta1
kind: DataImportCron
metadata:
  annotations:
    cdi.kubevirt.io/storage.bind.immediate.requested: "true"
  labels:
    template.kubevirt.io/architecture: "amd64"
    cdi.kubevirt.io/storage.import.datasource-name: "centos-stream9"
  name: centos-stream9-image-cron-amd64
spec:
  garbageCollect: Outdated
  managedDataSource: centos-stream9-amd64
  schedule: 0 */12 * * *
  template:
    spec:
      source:
        registry:
          url: docker://quay.io/containerdisks/centos-stream:9
          platform:
            architecture: "amd64"
      storage:
        resources:
          requests:
            storage: 10Gi
```

The `DataImportCron` CRs for `arm64` and `s390x` will be similar, differing only in their names, the value of the
`spec.template.spec.source.registry.platform.architecture` field, and the value of the
`template.kubevirt.io/architecture` label.
</div>

For backward compatibility, if the `ssp.kubevirt.io/dict.architectures` annotation is absent from a
`DataImportCronTemplate` object, SSP will maintain the current behavior by creating a single `DataImportCron`

> CR using the template's name (with no suffix), and without the new field and labels.

Reviewer 11: I've thought more about backward compatibility of SSP and the ability to disable this feature if needed. What do you think of this solution?

The new field in SSP spec `.spec.cluster` will be a pointer. 

If it is `nil`, SSP will behave as before this feature. It will create VM templates only for one architecture, `DataSource` names will not have architecture specific suffix, and it will not create multiple `DataImportCrons` form one `DataImportCronTemplate`, even if it has the `ssp.kubevirt.io/dict.architectures` annotation.

If  `.spec.cluster` is not `nil`:
- SSP will create multiple `DataSources`, one for each arch, plus the compatibility one that points to another `DataSource`.
- If a `DataImportCronTemplate` does not have the annotation and:
  - is used by a VM template -> SSP will create one `DataImportCron`, that will manage the `DataSource` for the same architecture as is the control plane. 
  - is not used by VM template -> SSP will behave as before this feature
- If there is a golden image `PVC` that was created by the cluster admin, SSP will assume that it contains image for control plane architecture. SSP will set the `DataSource` for this arch to point to this `PVC`, and it will show an alert to notify the user.

> CR using the template's name (with no suffix), and without the new field and labels.

Reviewer 11: We should not break existing functionality. If the admin has added a custom DataImportCron template, then it should continue working after this enhancement is implemented.

condition on the corresponding `DataImportCron` CR.

If the `spec.template.spec.source.registry.platform.architecture` field is not populated, the cdi-importer will pull the 
image without specifying an architecture, maintaining the current behavior.

### Cloning the Image into a Virtual Machine's PVC
In order to support the new "pointer" `DataSource` CR, The cdi-cloner, when populating the virtual-machine PVC, will use
the `status.source` field to find the right `VolumeSnapshot` or `PVC`, rather than the value of the `spec.source` field.

### Creating a Virtual Machine Templates
SSP will deploy all templates supported in the cluster, corresponding to the architectures listed in the new
`spec.cluster.workloadArchitectures` `SSP` CR's field. SSP dynamically add or remove templates if the architecture list
in the `SSP` CR changes.

For each architecture-specific template, SSP will set the `DATA_SOURCE_NAME` to the name of the
architecture-specific `DataSource`. The `DataSource` can be identified using its `template.kubevirt.io/architecture` and
the `cdi.kubevirt.io/storage.import.datasource-name` labels.

> 

Reviewer 11: This will break the edge case, when the admin has manually created a new golden image PVC, and the default `DataSource` points to it. Then after update to this proposal, the `DATA_SOURCE_NAME` will point to an architecture specific `DataSource`, which would not point to the existing PVC. 

We could document it, and inform the admin that the PVC has to be renamed to have an arch specific suffix.

`DataImportCronTemplate` objects to the new design at their own pace.
Additionally, it allows for independent implementation of components. For instance, if SSP implements this
design before the HCO, it will not introduce breaking changes.

If a `DataImportCronTemplate` object contains the `ssp.kubevirt.io/dict.architectures` annotation, but
a `DataImportCron` exists with the same name as the `DataImportCronTemplate`, and lacks the
`template.kubevirt.io/architecture` label, SSP will remove it. This behavior addresses two scenarios:

1. Cleaning up resources in the previous format during an upgrade from an older HCO version that does not implement this
   feature.
2. Cleaning up after a user adds the `cdi.kubevirt.io/storage.import.architectures` annotation to their custom
   `DataImportCronTemplate` object.

### Creating and Updating the `DataSource` Custom Resources
> **note**: The `DataSource` CR can be created by either CDI or SSP.

When creating or updating a `DataSource` CR, the `template.kubevirt.io/architecture` and the
`cdi.kubevirt.io/storage.import.datasource-name` labels must be propagated from the corresponding `DataImportCron` CR.

> When reconciling the `DataSource` CR, CDI will add the architecture to the `instancetype.kubevirt.io/default-preference`

Reviewer 5: Technically CDI already does this by either inspecting the image during Pod import and or propagating all of the related environment variables of the used container image in the event of a Node import. The thing is that we don't really have control over this. A user can populate the value however they see fit. For example, the official container disk image for Fedora which is multi-arch sets the value amd64 to "fedora" and "feodra.<arch>" for other architectures (i.e., arm64, s390x). Personally I do believe this is intended behavior, but this should be open for debate.

Since in the case of an architecture import the user explicitly states the desired architecture, we could acknowledge that intent by prefixing and or suffixing the required architecture. I personally, however, believe that we shouldn't.

Author: I would limit it to the pre-defined images, but CDI know nothing about them.

Reviewer 5: > I would limit it to the pre-defined images, but CDI knows nothing about them.
Indeed, that should be probably be done via DIC template label propagation.

Reviewer 2: The idea was to attach these labels to containerdisks and let CDI propagate these labels to imported PVCs. I'd say they should not be overwritten by CDI again. Can't we just drop this?

Author: I'm not deeply familiar with this specific logic. This is per request of @codingben.

@codingben - could you please answer here?

Reviewer 2: I read this again, so this is not a new feature? This should already work.

label, to be in the format of `<os>.<version>.<arch>`, e.g. `centos.stream9.amd64`.

#### Preserving the existing `DataSource`
The users may have an existing scripts or tools that are referencing to a specific `DataSource` CR; e.g. scripts to
create a specific type of virtual machine. These scripts are expected to keep working even after implementing the
heterogeneous cluster support.

To achieve that, CDI API will introduce the new "pointer" DataSource, by adding the `spec.source.dataSource` filed to
the DataSource API. 

SSP will create the architecture agnostic `DataSource` CR with no architecture label, and with the original name, 
as listed in the `spec.managedDataSource` field in the `DataImportCronTemplate` object (without architecture suffix). 
This `DataSource` CR point to the default `DataSource` CR, as described below.

For each image (represented by the same `DataImportCronTemplate` object in the SSP CR), SSP will select one of the
`DataSurce` CRs to be the default one. If there is only one supported architecture, this will be the default one.
If there are more than one supported architecture, the default one will be the one with the same architecture as the

> control-plane architecture, listed in the `spec.cluster.controlPlanArchitectures` field in the `SSP` CR.

Author: fixed in 3 places

`DataImportCron` CR, if provided.

The importer will pull the image corresponding to the specified architecture and loads it into the `VolumeSnapshot`.
1. if the `pullMethod` is `pod`, then the cdi-importer will use the platform field to pull the architecture specific
   image.
   > **Note**: when the cdi-importer is trying to pull an architecture specific image in a registry, it will set the
   > `os` field of the image platform to `linux`. The platform reflecting the runtime environment rather than the image
   > contents (i.e., VM images in KubeVirt may represent different operating systems).

2. when the `pullMethod` is `node`, the cdi-controller creates a pod with an init container with the required image
   letting the node's CRI to perform the actual pull. In this case the CRI will pull the image matches to its own 
   architecture. the cdi-controller will make sure that the pod will run on a node with the required architecture by
   modifying the pod's `nodeSelector` or `nodeAffinity` fields, to force scheduling on a node with the
   `kubernetes.io/arch` label, matching the architecture specified in the
   `spec.template.spec.source.registry.platform.architecture` field.

>    if there is no available node with the required architecture, the cdi-controller will report this failure in a

Author: @Acedus - setting the status does not fail the creation. It just a way to let the user know what's happening. I expect from CDI to keep trying, and of course, when a new node is added with the right architecture, the condition should be removed.

>    if there is no available node with the required architecture, the cdi-controller will report this failure in a

Reviewer 5: Just to make sure, the Pod condition and emitted event of scheduling failure is insufficient in that case?

>    if there is no available node with the required architecture, the cdi-controller will report this failure in a

Author: IMHO, the pod condition is not enough. First, I'm not sure what will happen to the pod in this case - will it gone, or stay as non-scheduled pod?

And then, the user does not have to know the implementation details, and where to search for errors. We need to make the error message accessible for the user in a higher level. I think the DIC level is enough.

>    if there is no available node with the required architecture, the cdi-controller will report this failure in a

Reviewer 5: > IMHO, the pod condition is not enough. First, I'm not sure what will happen to the pod in this case - will it gone, or stay as non-scheduled pod?

It will remain as pending.

> And then, the user does not have to know the implementation details, and where to search for errors. We need to make the error message accessible for the user in a higher level. I think the DIC level is enough.

Sure, that seems reasonable enough.

build time.

### Maintaining a List of Supported Architectures
The HCO will be enhanced to track the architectures of workload and control-plane nodes in the cluster. It will maintain
two distinct lists of these architectures, derived from the `status.nodeInfo.architecture` field of each node, one for 
the control-plane nodes and one for the workload nodes.

These lists will be dynamically updated as nodes are added or removed from the cluster.

By default, the HCO will identify workload nodes as those labeled with `node-role.kubernetes.io/worker`. If the
`spec.workloads.nodePlacement` field in the `HyperConverged` CR is defined, HCO will use this field to determine the
workload nodes instead.

HCO will identify control-plane nodes as those labeled with `node-role.kubernetes.io/control-plane`.

>**Note**: The control-plane architecture list is expected to always be with only one architecture, as the control-plane
> nodes are expected to be homogeneous. HCO will not force this, though.

### Reconciling the `SSP` Custom Resource
The following changes in HCO behavior will be applied if:

> 1. The new `enableMultiArchCommonBootImageImport` feature gate is enabled in the `HyperConverged` CR. This feature gate

Reviewer 2: ```suggestion
1. The new `EnableMultiArchCommonBootImageImport` feature gate is enabled in the `HyperConverged` CR. This feature gate
```

Don't these typically begin with a capital letter?

Author: HCO FGs are fields, rather than KubeVirt list of strings. yaml standard for filed names is to start with a low case .

   is disabled by default.
2. The cluster is not a single node cluster.

When reconciling the `SSP` CR, HCO will:

1. Populate the `spec.cluster.workloadArchitectures` filed in the `SSP` CR with the architectures of the workload nodes
   currently present in the cluster.

2. Populate the `spec.cluster.controlPlaneArchitectures` filed in the `SSP` CR with the architectures of the
   control-plane nodes currently present in the cluster.
   > **Note**: The control-plane architecture list is expected to always be with only one architecture.

3. The pre-prepared DataImportCronTemplate files in the HCO image, will already contain the 
   `ssp.kubevirt.io/dict.architectures` annotation for each `DataImportCronTemplate`. When adding the a
   `DataImportCronTemplate` object to the `SSP` CR, HCO will clean the architectures that are not present in cluster
    from the annotation.
   

>    If there are no architectures that supported by the image and present in the cluster, HCO will not add this

Reviewer 2: ```suggestion
   If there are no architectures that are supported by the image and present in the cluster, HCO will not add this
```

   image.
   > **Note**: when the cdi-importer is trying to pull an architecture specific image in a registry, it will set the
   > `os` field of the image platform to `linux`. The platform reflecting the runtime environment rather than the image
   > contents (i.e., VM images in KubeVirt may represent different operating systems).

2. when the `pullMethod` is `node`, the cdi-controller creates a pod with an init container with the required image
   letting the node's CRI to perform the actual pull. In this case the CRI will pull the image matches to its own 
   architecture. the cdi-controller will make sure that the pod will run on a node with the required architecture by
   modifying the pod's `nodeSelector` or `nodeAffinity` fields, to force scheduling on a node with the
   `kubernetes.io/arch` label, matching the architecture specified in the
   `spec.template.spec.source.registry.platform.architecture` field.

   if there is no available node with the required architecture, the cdi-controller will report this failure in a
   condition on the corresponding `DataImportCron` CR.

> If the image is not a manifest or does not include the specified architecture, the CDI will report this failure in a

Reviewer 2: What if it is a single image (no manifest) but the correct arch?

Author: Then its DICT must not have the `ssp.kubevirt.io/dict.architectures` annotation.

condition on the corresponding `DataImportCron` CR.

If the `spec.template.spec.source.registry.platform.architecture` field is not populated, the cdi-importer will pull the 
image without specifying an architecture, maintaining the current behavior.

### Cloning the Image into a Virtual Machine's PVC
In order to support the new "pointer" `DataSource` CR, The cdi-cloner, when populating the virtual-machine PVC, will use
the `status.source` field to find the right `VolumeSnapshot` or `PVC`, rather than the value of the `spec.source` field.

### Creating a Virtual Machine Templates
SSP will deploy all templates supported in the cluster, corresponding to the architectures listed in the new
`spec.cluster.workloadArchitectures` `SSP` CR's field. SSP dynamically add or remove templates if the architecture list
in the `SSP` CR changes.

For each architecture-specific template, SSP will set the `DATA_SOURCE_NAME` to the name of the
architecture-specific `DataSource`. The `DataSource` can be identified using its `template.kubevirt.io/architecture` and
the `cdi.kubevirt.io/storage.import.datasource-name` labels.

### Creating a Virtual Machine from a Template

> Assuming the user knows the architecture of the VM they want to create, the kubevirt-plugin (UI) will prompt the user to

Reviewer 2: Do we need to make adjustments for VirtualMachinePreferences in the UI?

Author: Please suggest the require change.

select the desired architecture. Once selected, the UI will display templates corresponding to the chosen architecture.

While the above assumption is valid, this approach contrasts with practices observed in other virtualization solutions.
For instance, AWS integrates architecture selection directly into the template selection UI:

![aws-select-arch-ui.png](aws-select-arch-ui.png)

## API Changes

### CDI API Changes
The following API changes are required:
The `DataVolumeSourceRegistry` type in the CDI API will be extended with the new `platform` field, with the new type of
`PlatformOptions`; e.g.

```go
type DataVolumeSourceRegistry struct {
//Platform describes the minimum runtime requirements of the image
// +optional
Platform *PlatformOptions `json:"platform,omitempty"`
}
```

```go
type PlatformOptions struct {
//Architecture specifies the image target CPU architecture
// +optional
Architecture string `json:"architecture,omitempty"`
// +optional
//Variant specifies the image target CPU architecture variant
Variant string `json:"variant,omitempty"`
}
```

Add the "pointer" `DataSource` type, by adding the `spec.source.dataSource` field to the `DataSource` API:
```go

> type DataSourceSource struct {

Author: @Acedus , WDYT?

> type DataSourceSource struct {

Author: @Acedus started the implementation, here: https://github.com/kubevirt/containerized-data-importer/pull/3760

I OK with this change.

> type DataSourceSource struct {

Reviewer 5: Personally I'm indifferent to it but during initial POC I did find it more convenient to refer to it as a reference/pointer/link so I tend to concur.

> type DataSourceSource struct {

Author: cool. So I'll fix the VEP. @Acedus please fix your PR to use "DataSourceRefSource". But I think the struct name is less important, but the field that uses it, which is part of the API and exposed in the yaml. Let's change that one too:

```go
type DataSourceRefSource struct {
	// The namespace of the source DataSource
	Namespace *string `json:"namespace"`
	// The name of the source DataSource
	Name *string `json:"name"`
}

type DataSourceSource struct {
	...
	// +optional
	dataSourceRef *DataSourceRefSource `json:"dataSource,omitempty"`
}
```

@Acedus - WDYT?

> type DataSourceSource struct {

Reviewer 5: Note that it should be "DataVolumeSourceRefDataSource`, as for the struct embed I don't see a reason to change it to `DataSourceRef`, I believe it's implied.

> type DataSourceSource struct {

Author: Oh, I understand my confusion. So the original comment here was about the existing type `DataSourceSource`. This type is the type of the existing  DataSource's `spec.source` field. We just adding a new sub field here, and its type.

bottom line: we'll add the `dataSource` field to the existing `DataSourceSource`:
```go
ype DataSourceSource struct {
	// +optional
	PVC *DataVolumeSourcePVC `json:"pvc,omitempty"`
	// +optional
	Snapshot *DataVolumeSourceSnapshot `json:"snapshot,omitempty"`
	// +optional
	DataSource *DataSourceRefSourceDataSource `json:"dataSource,omitempty"` /// <<== this is the new addition
}
```
And the new `DataSourceRefSourceDataSource` type:
```go
type DataSourceRefSourceDataSource struct {
	// The namespace of the source DataSource
	Namespace *string `json:"namespace"`
	// The name of the source DataSource
	Name *string `json:"name"`
}
```
